Approved/Initializers.md

```diff
@@ -0,0 +1,175 @@
+Custom initializers take an argument of an arbitrary type, provided it does not contain items of allocatable types. This restriction makes it possible to safely allocate and de-allocate temporary qubits when nesting initializer expressions (see the paragraph further below). 
+
+### *Creating initializers on the fly by elevating operations*:   
+Since it is a common scenario to initialize qubits to a certain state, this proposal includes the means to construct initializers on the fly that do just that. There are two distinct cases that are common. 
+1. One is that an operation should be invoked on the allocated qubits after initialization, followed by further computations before the qubits are measured and released. In this case, it is not necessary or desirable that the operation performed upon initialization is un-computed upon release. 
+2. In the second case, the computations following the initialization lead to the qubits being in the same state as they were initialized. In this case, the adjoint of the operation used for initialization needs to be applied upon releasing the qubits. 
+
+The proposal is to syntactically distinguish the two cases. The table below contains examples for both cases. In the first case, the initializer consists of the keyword `init`, followed by the initializer argument tuple, the keyword `then` and an operation-valued expression. The initializer argument tuple may be omitted if the required argument is of type `Unit`. The second case follows the same pattern with the exception that instead of the keyword `then`, then keyword `within` should be used, and the operation-valued expression needs to support the `Adjoint` functor.   
+In both cases, the elevated operation has to return `Unit`.
+The allocated value matches the argument of the elevated operation and will be bound to the symbol(s) specified on the left hand side of the allocation statement.  
+
+### *Initializer expressions and nesting of initializers*:    
+An initializer expressions consists either of a call to an initializer, or of a tuple of multiple calls. ...
+
+### Overview over different kinds of initializer expressions
+
+Initializers for built-in types:
+| Built-in Initializer | Description | Allocated Type | Initializer Argument Type | Initializer Expression | 
+| --- | --- | --- | --- | --- |
+| Qubit | allocates a single qubit | `Qubit` | `Unit` | `Qubit()` |
+| Qubits | allocates a 1D array of qubits | `Qubit[]` | `Int` | `Qubits(4)` |
+| Qubits | allocates a [2D array](https://github.com/microsoft/qsharp-language/issues/39) of qubits | `Qubit[,]` | `(Int, Int)` | `Qubits(2,4)` |
```

Comment:
Effectively, it seems this is a kind of overloading, in that `Qubits` can either have type `Int ~> Qubit[]` or `(Int, Int) ~> Qubit[,]`. Is that represented by a type parameter to `Qubits`?

Reply:
The overloading on tuple arity seems weird to me - it's compiler magic that can't be expressed in the language itself. It would not be possible to write a custom initializer that does this. A type parameter would allow any type, not just tuples of `Int`s.